Legacy File Redux

rockwood opened the pull request:
This implements the proposal in #81.

Now, plugins register the legacy filenames they support. ASDF searches for both legacy and .tool-version files simultaneously.

Users with plugins that only support the old legacy file api will see a warning message with instructions to update their plugin.

Stratus3D commented on test/utils.bats:
```diff
@@ -38,8 +41,100 @@ teardown() {
+  run check_if_plugin_exists "dummy"
+  [ "$status" -eq 0 ]
+  [ "$output" = "" ]
+}
```
Duplicate of the test above.

Stratus3D commented:
This is awesome! I've had this PR open in a tab for the last week and just got around to reading the code. I think this is exactly what we need. 

We might want to implement the new legacy version scripts in the plugins that contain the get_version_from_legacy_file script before merging this though, since the updated plugins will continue to work with the current version of asdf, but once this is merged it won't work with old plugins (other than the warning).

rockwood commented:
@Stratus3D Good catch. I removed the duplicate test and rebased master. 

I just opened PRs for the two plugins that currently support the legacy file api.

https://github.com/asdf-vm/asdf-ruby/pull/17
https://github.com/tuvistavie/asdf-python/pull/4

Once those get merged, we should be good to merge this.

rockwood commented:
I'm closing this for the time being due to a bug with conflicting local and global versions. I'll reopen when I get that fixed

rockwood commented:
Ok, all fixed.

Stratus3D commented:
Great! I just merged the PR for the Ruby plugin, and it looks like the Python one has already been merged. 

I would like to have one more maintainer look at this before merging. @asdf-vm/maintainers what do you all think about these improvements to the legacy version file feature?

tuvistavie commented on docs/creating-plugins.md:
```diff
@@ -50,9 +50,19 @@ Setup the env to run the binaries in the package.
+
+#### bin/parse-legacy-file
+
+This can be used to further parse the legacy file found by asdf. If `parse-legacy-file` isn't implemented, asdf will simply cat the file to determine the version.
```
Maybe we could add

> The script will be passed the file path as its first argument.

or something like this to make it clear that we are passing the file path, and not its content.
I was not totally sure which one was passed before checking the implementation.

tuvistavie commented:
It looks great, thank you very much! :smile:
I just left a little comment about the README, let me know what you think!

rockwood commented on docs/creating-plugins.md:
```diff
@@ -50,9 +50,19 @@ Setup the env to run the binaries in the package.
+
+#### bin/parse-legacy-file
+
+This can be used to further parse the legacy file found by asdf. If `parse-legacy-file` isn't implemented, asdf will simply cat the file to determine the version. The script will be passed the file path as its first argument.
```
@tuvistavie Good call.
